README.md

```diff
@@ -1,53 +1,251 @@
+tells the dataset, that it should keep all records within 10 of the
+current read offset loaded. That's why it fetched the first two
+pages. Now logicially, our dataset looks like this:
+
+> Note: `*` indicates that the record is pending.
+
+
+```
+           Read
+          Offset
+             ┃
+             ┃
+<──────────Load Horizon──────────>
+             ┃
+             ▼
+             ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+              * * * * * * * * * *
+             ◇ ─ ─ ─ ─ ◇─ ─ ─ ─  ┘
+             │         │
+             p0        p1
+```
+
+At some point, the request for the first page resolves. At that point,
+the dataset will emit a new state with the resolved records. That
+state will still contain the pending records as well as the freshly
+loaded.
+
+
+```javascript
+//get a record off the first page
+record = state.get(3);
+record.isResolved //=> true
+record.content //=> 3
+
+//get a record off the second page
+record = state.get(7)
+record.isPending //=> true
+```
+
+Another interesting thing that happened here is that the length of the
+dataset has also changed.
+
+```javascript
+state.length //=> 30
+```
+
+This has to do with the `stats` parameter that is passed into the
+fetch function. This value allows the fetch function to optionally
+specify the total extent of the dataset if that information is
+available. This can be useful when rendering native scrollbars or
+other UI elements that indicate the overall length of a list. If
+`stats` are never updated, then the dataset will just expand
+indefinitely. Now our state looks like this:
+
+> Note `x` indicates that the record is not yet requested
+
+```
+            Read
+           Offset
+              ┃
+              ┃
+<──────────Load Horizon──────────>
+              ┃
+              ▼
+              ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+               0 1 2 3 4 * * * * * xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx│
+              ◇ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ─ ─ ─◇─ ─ ─ ─ ─ ─ ─ ◇ ─ ─ ─ ─ ─ ─ ─
+              │         │         │              │              │
+             p0        p1        p2             p3             p4
+
+```
+
+We have records 0-4, whilst records 5-9 are in flight, and records
+10-25 have yet to be requested.
+
+```javascript
+//from the last page (p4)
+record = state.get(23);
+record.isRequested //=> false
+record.isPending //=> false
+record.content //> null
+```
+
+Let's say we want to move the read head to offset 2 with a call to
+`dataset.setReadOffset(2)`. This will immediately emit a new state that
+looks like this:
+
+```
+                 Read
+                Offset
+                   ┃
+                   ┃
+     <──────────Load Horizon──────────>
+                   ┃
+                   ▼
+              ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+               0 1 2 3 4 * * * * * ** ** ** ** ** xx xx xx xx xx xx xx xx xx xx│
```

Comment:
hah, those are numbers with two digits. I wanted to make sure that the length of the dataset didn't change overall. Not sure what to do there.